### PR TITLE
SI-7733 reflective packages now more consistent with scalac 

### PR DESCRIPTION
**Pull request by xeno-by**
Previously PackageScopes from scala.reflect ignored all classes that
had $'s in non-rightmost positions in their names.

Unfortunately this behaviour is inconsistent with how scalac does things,
and I reconciled this as usual, by pulling corresponding logic into
scala-reflect.jar and sharing it between runtime reflection and compiler.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @dragos @jhaberstro

**Comment by xeno-by** [comment]
@adriaanm @retronym Please consider merging this pull request into 2.10.3, as it appears to be very important for the Scala IDE team. I guess, @dragos will be able provide more details.

**Comment by dragos** [comment]
A bit of context: while prototyping "conditional breakpoint" in the debugger, we turned to the toolbox compiler: compile the condition once, evaluate it every time the line is executed. However, if the breakpoint is inside a closure, there was no way to bind `this` to a type that contains the captured environment (`..$anonfun$1`), so the conditional expression was very limited.

It would be great to have this in the next release, but it's not a feature that we will release very soon either.

**Comment by xeno-by** [comment]
Could someone elaborate on what do the build results mean here? Why's the build failed? The console output page doesn't provide an explanation. https://scala-webapps.epfl.ch/jenkins/job/pr-scala/915/

**Comment by retronym** [comment]
Same problem for me here: https://github.com/scala/scala/pull/2771

**Comment by adriaanm** [comment]
We've switched to using the build flow plugin, which was a big mistake. It's incredibly buggy. I don't know why the output disappears, but the console output is still there, luckily.

I tried a lot of different ways of setting this up on jenkins, made staging versions of the jobs and ran them under as much load as the bot generates before switching to it in production. When all seemed to be going well, I switched the bot to the pr-scala job, but new IDE failures appeared under even higher load/randomness, so I tried to make the flow ignore the result of the ide validation, and ran into more bugs with the flow plugin. I've now resorted to a solution that should work (make the IDE job always succeed -- suboptimal for discovering failures, but the best I can do...)

**Comment by xeno-by** [comment]
@adriaanm I'm still unsure whether the failure was spurious or not. Kitten points to https://scala-webapps.epfl.ch/jenkins/job/pr-scala/915/console, but it doesn't provide much info.

**Comment by adriaanm** [comment]
It's spurious -- the jenkins build flow plugin is broken IMO. I've resolved the issue as explained above and at #2771.

PLS REBUILD ALL

**Comment by xeno-by** [comment]
Thank you!

**Comment by xeno-by** [comment]
Allright looks good to merge to Jenkins. Can any human please join him?

**Comment by dragos** [comment]
LGTM.

File: src/compiler/scala/tools/nsc/util/ClassPath.scala [hunk 1]
```diff
@@ -100,7 +101,8 @@ object ClassPath {
   }
 
   /** A useful name filter. */
-  def isTraitImplementation(name: String) = name endsWith "$class.class"
+  @deprecated("use scala.reflect.runtime.ReflectionUtils.isTraitImplementation instead", "2.11.0")
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/util/ClassPath.scala]: This is wrong. You are about to merge changes to 2.10.x branch which mention deprecation in "2.11".

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/util/ClassPath.scala]: That's right. Should I mention 2.10.3 here or should I submit deprecation as a separate request for 2.11?

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/util/ClassPath.scala]: We can't deprecate members in minor releases. Please remove it from here and submit a separate PR against master. Thanks.

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/util/ClassPath.scala]: Done and done: https://github.com/scala/scala/pull/2825.
